corehq/apps/locations/util.py

```diff
@@ -241,68 +274,39 @@ def coax_boolean(value):
             return value
 
         rows = []
-        for lt in location_types:
-            type_row = {
-                headers['code']: lt.code,
-                headers['name']: lt.name,
-                headers['parent_code']: foreign_code(lt, 'parent_type'),
-                headers['do_delete']: '',
-                headers['shares_cases']: coax_boolean(lt.shares_cases),
-                headers['view_descendants']: coax_boolean(lt.view_descendants),
-            }
-            rows.append(dict(type_row))
-
-        return ('types', {
-            'headers': [headers[header] for header in ['code', 'name', 'parent_code', 'do_delete',
-                        'shares_cases', 'view_descendants']],
-            'rows': rows
-        })
-
-    def get_export_dict(self):
-        location_types = self.domain_obj.location_types
-        sheets = [self.type_sheet(location_types)]
-        sheets.extend([
-            self._loc_type_dict(loc_type)
-            for loc_type in location_types
-        ])
-        return sheets
-
+        for lt in self.location_types:
+            rows.append([
+                lt.code,
+                lt.name,
+                foreign_code(lt, 'parent_type'),
+                '',  # do_delete
+                coax_boolean(lt.shares_cases),
+                coax_boolean(lt.view_descendants),
+            ])
 
-def dump_locations(domain, download_id, include_consumption=False):
-    exporter = LocationExporter(domain, include_consumption=include_consumption)
-    use_transfer = settings.SHARED_DRIVE_CONF.transfer_enabled
-    filename = '{}_locations.xlsx'.format(domain)
-    file_path = write_to_file(exporter.get_export_dict(), filename, use_transfer)
+        writer.write([('types', rows)])
 
-    expose_download(use_transfer, file_path, filename, download_id, 'xlsx')
 
+def dump_locations(domain, download_id, include_consumption=False, task=None):
+    exporter = LocationExporter(domain, include_consumption=include_consumption, async_task=task)
 
-def write_to_file(locations, filename, use_transfer):
-    """
-    locations = [
-        ('loc_type1', {
-             'headers': ['header1', 'header2', ...]
-             'rows': [
-                 {
-                     'header1': val1
-                     'header2': val2
-                 },
-                 {...},
-             ]
-        })
-    ]
-    """
-    outfile = get_download_file_path(use_transfer, filename)
+    fd, path = tempfile.mkstemp()
     writer = Excel2007ExportWriter()
-    header_table = [(tab_name, [tab['headers']]) for tab_name, tab in locations]
-    writer.open(header_table=header_table, file=outfile)
-    for tab_name, tab in locations:
-        headers = tab['headers']
-        tab_rows = [[row.get(header, '') for header in headers]
-                    for row in tab['rows']]
-        writer.write([(tab_name, tab_rows)])
+    writer.open(header_table=exporter.get_headers(), file=path)
+    exporter.write_data(writer)
```

Comment:
Think it's worth sticking this in a try/finally block so the writer is always closed?  It also wouldn't be hard to stick `__enter__` and `__exit__` on `Excel2007ExportWriter` so we can use `with`, but that's probably out of the scope of this PR.  I'll make a note to do that at some point if you don't get to it here.

Reply:
I did think about that (making `Excel2007ExportWriter` a context manager), not sure why I didn't as I don't think it's a big lift.

Reply:
2d0fecb